File: src/main/kotlin/com/github/bestheroz/standard/common/mybatis/handler/GenericEnumTypeHandler.java
package com.github.bestheroz.standard.common.mybatis.handler;

import com.github.bestheroz.standard.common.enums.ValueEnum;
import org.apache.ibatis.type.BaseTypeHandler;
import org.apache.ibatis.type.JdbcType;

import java.sql.CallableStatement;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class GenericEnumTypeHandler<E extends ValueEnum> extends BaseTypeHandler<E> {

  private final Class<E> type;

  public GenericEnumTypeHandler(Class<E> type) {
    if (type == null) throw new IllegalArgumentException("Type argument cannot be null");
    this.type = type;
  }

  @Override
  public void setNonNullParameter(PreparedStatement ps, int i, E parameter, JdbcType jdbcType)
      throws SQLException {
    if (parameter == null) {
      ps.setNull(i, jdbcType.TYPE_CODE);
    } else {
      ps.setString(i, parameter.value);
    }
  }

  @Override
  public E getNullableResult(ResultSet rs, String columnName) throws SQLException {
    String value = rs.getString(columnName);
    return getEnum(value);
  }

  @Override
  public E getNullableResult(ResultSet rs, int columnIndex) throws SQLException {
    String value = rs.getString(columnIndex);
    return getEnum(value);
  }

  @Override
  public E getNullableResult(CallableStatement cs, int columnIndex) throws SQLException {
    String value = cs.getString(columnIndex);
    return getEnum(value);
  }

  private E getEnum(String value) throws SQLException {
    if (value == null) return null;
    try {
      for (E enumConstant : type.getEnumConstants()) {
        if (enumConstant.value.equals(value)) {
          return enumConstant;
        }
      }
      throw new IllegalArgumentException("Unknown enum value: " + value);
    } catch (Exception e) {
      throw new SQLException(
          "Cannot convert " + value + " to " + type.getSimpleName() + " by name.", e);
    }
  }
}

File: src/main/kotlin/com/github/bestheroz/standard/common/mybatis/handler/GenericEnumTypeHandler.kt
package com.github.bestheroz.standard.common.mybatis.handler

import io.github.bestheroz.mybatis.type.ValueEnum
import org.apache.ibatis.type.BaseTypeHandler
import org.apache.ibatis.type.JdbcType
import java.sql.CallableStatement
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.SQLException

class GenericEnumTypeHandler<E : ValueEnum>(
    private val type: Class<E>?,
) : BaseTypeHandler<E>() {
    init {
        requireNotNull(type) { "Type argument cannot be null" }
    }

    override fun setNonNullParameter(
        ps: PreparedStatement,
        i: Int,
        parameter: E?,
        jdbcType: JdbcType?,
    ) {
        if (parameter == null) {
            // jdbcType?.TYPE_CODE가 null이면 기본값 0 지정
            ps.setNull(i, jdbcType?.TYPE_CODE ?: 0)
        } else {
            ps.setString(i, parameter.value)
        }
    }

    override fun getNullableResult(
        rs: ResultSet,
        columnName: String?,
    ): E? = getEnum(rs.getString(columnName))

    override fun getNullableResult(
        rs: ResultSet,
        columnIndex: Int,
    ): E? = getEnum(rs.getString(columnIndex))

    override fun getNullableResult(
        cs: CallableStatement,
        columnIndex: Int,
    ): E? = getEnum(cs.getString(columnIndex))

    @Throws(SQLException::class)
    private fun getEnum(value: String?): E? {
        if (value == null) {
            return null
        }
        return try {
            for (enumConstant in checkNotNull(type).enumConstants) {
                if (enumConstant.value == value) {
                    return enumConstant
                }
            }
            throw IllegalArgumentException("Unknown enum value: $value")
        } catch (e: Exception) {
            throw SQLException("Cannot convert $value to ${type?.simpleName} by name.", e)
        }
    }
}
